### PR TITLE
CSS-3453: Add scaling functionality

### PR DESCRIPTION
**Pull request by kelkawi-a**
This PR introduces the option of deploying the [Temporal Server](https://docs.temporal.io/clusters) as its four separate services (frontend, history, matching and worker). As these services can now be deployed as four different applications, they can be scaled independently according to their needs. The changes made are as follows:

- The charm now opens and closes ports based on the services defined in the config.
- Unit and integration tests have been added to test the scaling functionality (Required upgrading ops to 2.2.0).
- Documentation has been added to describe how to deploy the Temporal Cluster.
- Separated integration tests workflow into charm and scaling tests.

### CODE REVIEW

File: documentation/how-to/scaling.md [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+The Temporal server charm is designed such that each service can be deployed as
+a separate application which can then be connected together by integrating with
+the same database. These services can then be scaled according to your needs.
+
+To deploy the services separately in a scalable way, you must deploy the
+application as follows:
+
+```bash
+juju deploy temporal-k8s --config services="frontend"
```
Comment (mertalpt) [inline documentation/how-to/scaling.md:30]: I thought we only had the UI and worker charms for Temporal. What is `temporal-k8s-history` for example? I am asking for myself.

**Review by mertalpt — APPROVED** [review]
LGTM.

File: lib/charms/loki_k8s/v0/loki_push_api.py [hunk 2]
```diff
@@ -1762,7 +1766,7 @@ class LogProxyConsumer(ConsumerBase):
     def __init__(
         self,
         charm,
-        log_files: Optional[list] = None,
+        log_files: Optional[Union[List[str], str]] = None,
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply lib/charms/loki_k8s/v0/loki_push_api.py:1769]: This is part of the loki-k8s charm library update, we are simply importing the library to use it with our Temporal charm.

File: documentation/how-to/scaling.md (hunk 1 above)
Comment (kelkawi-a) [reply documentation/how-to/scaling.md:30]: Sorry for the confusion! I just updated the instructions as it should have been ```juju deploy temporal-k8s --config services="history" temporal-k8s-history```, meaning it will deploy the ```temporal-k8s``` charm with the name ```temporal-k8s-history``` and assign the services config value to be "history" only.

File: documentation/how-to/scaling.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+temporal-k8s-matching/0*  active    idle   10.1.232.61
+temporal-k8s-matching/1   active    idle   10.1.232.38
+temporal-k8s-worker/0*    active    idle   10.1.232.42
+temporal-k8s-worker/1     active    idle   10.1.232.23
+temporal-k8s/0*           active    idle   10.1.232.26
+temporal-k8s/1            active    idle   10.1.232.25
+```
+
+To confirm the four scaled services can reach other, you can run
+`juju run temporal-admin-k8s/0 tctl args="adm cl d"`, you should see the
```
Comment (kian99) [inline documentation/how-to/scaling.md:219]: Is this bit plus the output not just a repeat of the above? I.e. can we just say refer to above or something similar.

File: documentation/how-to/scaling.md (hunk 1 above)
Comment (kian99) [reply documentation/how-to/scaling.md:30]: Ah was wondering something similar, very cool the way you can use the same charm to spin up the individual components.

**Review by kian99 — APPROVED** [review]

File: documentation/how-to/scaling.md (hunk 3 above)
Comment (kelkawi-a) [reply documentation/how-to/scaling.md:219]: I could remove it, but I've included a second copy because I thought it would be useful to see how the output would look like when we have replicas (i.e. the way each service has two reachable units).

Comment (kian99) [reply documentation/how-to/scaling.md:219]: Gotcha, that's fine in that case.

File: src/charm.py [hunk 4]
```diff
@@ -33,7 +33,14 @@
 
 logger = logging.getLogger(__name__)
 
-SERVER_PORT = 7233
+FRONTEND_PORT_GRPC = 7233
+FRONTEND_PORT_HTTP = 6933
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/charm.py:37]: According to the Temporal documentation, these are for "membership-related communication".

File: src/charm.py [hunk 5]
```diff
@@ -344,6 +382,9 @@ def _update(self, event):
                 "VISIBILITY_PORT": visibility_conn["port"],
                 "VISIBILITY_USER": visibility_conn["user"],
                 "VISIBILITY_PSWD": visibility_conn["password"],
+                "TEMPORAL_BROADCAST_ADDRESS": str(self.model.get_binding("peer").network.bind_address),
+                # TODO(kelkawi-a): do not assume the app is always deployed with this name.
+                "PUBLIC_FRONTEND_ADDRESS": "temporal-k8s:7233",
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/charm.py:387]: As we are deploying multiple temporal-k8s charms with different names for different services, we want them all to point to the one named ```temporal-k8s``` which will be handling the frontend service.

File: documentation/how-to/scaling.md [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+The Temporal Server services can run independently or be grouped together into
+shared processes on one or more physical or virtual machines. For live
+(production) environments, we recommend that each service runs independently,
+because each one has different scaling requirements and troubleshooting becomes
+easier. The History, Matching, and Worker Services can scale horizontally within
+a Cluster. The Frontend Service scales differently than the others because it
+has no sharding or partitioning; it is just stateless.
+
+(The above was extracted from
+[Temporal's documentation](https://docs.temporal.io/clusters#temporal-server))
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply documentation/how-to/scaling.md]: Noted. Thanks!

**Review by gtato — APPROVED** [review]

File: src/charm.py [hunk 7]
```diff
@@ -308,6 +311,38 @@ def _validate(self):
         if not self._state.schema_ready:
             raise ValueError("admin:temporal relation: schema is not ready")
 
+    def _open_service_ports(self):
+        """Open the respective ports based on Temporal service."""
```
Comment (frankban) [inline src/charm.py:315]: Alternatively, and more concisely:
```python
SERVICE_PORTS = {
    "matching": {
        "grpc": 7235,
        "http": 6935, 
    },
    "worker": {
        "grpc": 7239,
        "http": 6939, 
    },
    ...
}

open_port = functools.partial(self.model.unit.open_port, protocol="tcp")
close_port = functools.partial(self.model.unit.close_port, protocol="tcp")

for service, ports in SERVICE_PORTS.items():
    if service in services:
        open_port(service["grpc"])
        open_port(service["http"])
    else:
        close_port(service["grpc"])
        close_port(service["http"])

```

File: src/charm.py [hunk 8]
```diff
@@ -344,6 +382,9 @@ def _update(self, event):
                 "VISIBILITY_PORT": visibility_conn["port"],
                 "VISIBILITY_USER": visibility_conn["user"],
                 "VISIBILITY_PSWD": visibility_conn["password"],
+                "TEMPORAL_BROADCAST_ADDRESS": str(self.model.get_binding("peer").network.bind_address),
+                # TODO(kelkawi-a): do not assume the app is always deployed with this name.
```
Comment (frankban) [inline src/charm.py:386]: Isn't getting the name dynamically simple enough?

**Review by frankban — APPROVED** [review]
I know this one is merged, and it looks good, but I had a couple of comments pending. Thanks!

File: src/charm.py (hunk 7 above)
Comment (kelkawi-a) [reply src/charm.py:315]: Thanks for the advice! Will include it in the next PR.

File: src/charm.py (hunk 8 above)
Comment (kelkawi-a) [reply src/charm.py:386]: So as it stands currently, the two services (matching and worker) must talk to the frontend service (which is deployed using the name ```temporal-k8s```). With a scaled deployment, the four services are deployed as four different applications (which are related and communicate at the database level, not at the juju level). Therefore, we would need to figure out a way of getting the ```temporal-k8s``` charm to communicate its name with the other two charms deployed for them to correctly communicate with it. 

TLDR: ```PUBLIC_FRONTEND_ADDRESS``` config is being set to ```temporal-k8s:7233``` to allow the matching and worker services to communicate with it.
